src/unify/unify.fun

```diff
@@ -0,0 +1,226 @@
+      val vars1 = Vars.fromListDistinct xs1 handle Vars.Duplicate => raise Pattern
+      val vars2 = Vars.fromListDistinct xs1 handle Vars.Duplicate => raise Pattern
+
+      val vl1 = ((List.map #2 rs1, List.map sort tms1), tau)
+      val vl2 = ((List.map #2 rs2, List.map sort tms2), tau)
+    in
+      if Syms.subset (syms1, syms2) andalso Vars.subset (vars1, vars2) then
+        let
+          val bnd = (us2, xs2) \ check (X1 $# (rs1, tms1), tau)
+        in
+          Metavar.Ctx.insert rho X2 @@ checkb (bnd, vl2)
+        end
+      else if Syms.subset (syms2, syms1) andalso Vars.subset (vars2, vars1) then
+        let
+          val bnd = (us1, xs1) \ check (X2 $# (rs2, tms2), tau)
+        in
+          Metavar.Ctx.insert rho X1 @@ checkb (bnd, vl1)
+        end
+      else
+        let
+          val X3 = freshMeta ()
+          val rs' = List.mapPartial (fn r as (O.P.VAR u, sigma) => if Syms.member syms2 u then SOME r else NONE | _ => NONE) rs1
+          val tms' = List.mapPartial (fn tm => case out tm of `x => if Vars.member vars2 x then SOME tm else NONE | _ => NONE) tms1
+          val tm = check (X3 $# (rs', tms'), tau)
+          val bnd1 = (us1, xs1) \ tm
+          val bnd2 = (us2, xs2) \ tm
+          val rho' = Metavar.Ctx.insert rho X2 @@ checkb (bnd2, vl2)
+        in
+          Metavar.Ctx.insert rho' X1 @@ checkb (bnd1, vl1)
+        end
+    end
+
+  fun unify_ (pvars, rho) (tm1, tm2) =
+    let
+      val tm1' = substMetaenv rho tm1
+      val tm2' = substMetaenv rho tm2
+      fun fail () = raise Unify (tm1', tm2')
+
+      val tau = sort tm1
+      val _ = if O.Ar.Vl.S.eq (tau, sort tm2) then () else raise Sort
+    in
+      case (out tm1', out tm2') of
+         (`x1, `x2) => if Var.eq (x1, x2) then rho else fail ()
+       | (theta1 $ bs1, theta2 $ bs2) =>
+         if O.eq Sym.eq (theta1, theta2) then
+           ListPair.foldlEq (fn (b1, b2, rho) => unifyb (pvars, rho) (b1, b2)) rho (bs1, bs2)
+         else
+           fail ()
+       | (X1 $# (rs1, tms1), X2 $# (rs2, tms2)) =>
+         (case (Metas.member pvars X1, Metas.member pvars X2) of
+            (true, true) =>
+            if Metavar.eq (X1, X2) then
+              (flexFlex1 rho (X1, tau, rs1, rs2, tms1, tms2) handle _ => fail ())
+            else
+              (flexFlex2 rho (X1, X2, tau, rs1, rs2, tms1, tms2) handle _ => fail ())
+          | (false, false) =>
+            if Metavar.eq (X1, X2) andalso ListPair.allEq (O.P.eq Sym.eq) (List.map #1 rs1, List.map #1 rs2) then
+              ListPair.foldlEq (fn (tm1, tm2, rho) => unify_ (pvars, rho) (tm1, tm2)) rho (tms1, tms2)
+            else
+              fail ()
+          | (true, false) => (flexRigid (pvars, rho) (X1, sort tm1', rs1, tms1, tm2') handle _ => fail ())
+          | (false, true) => (flexRigid (pvars, rho) (X2, sort tm2', rs2, tms2, tm1') handle _ => fail ()))
+       | (X $# (rs, tms), _) => (flexRigid (pvars, rho) (X, sort tm1', rs, tms, tm2') handle _ => fail ())
```

Comment:
This should be calling `flexRigid` only if `X` is in `pvars`. Otherwise, unification should fail. Same with the next clause.